get-credentials-gke/src/gkeClient.ts

```diff
@@ -0,0 +1,198 @@
+      }
+      jsonContent = JSON.parse(creds);
+      this.auth.jsonContent = jsonContent;
+    }
+    // Set project Id
+    if (!projectId && jsonContent && jsonContent.project_id) {
+      projectId = jsonContent.project_id;
+      core.info('Setting project Id from credentials');
+    } else if (!projectId && process.env.GCLOUD_PROJECT) {
+      projectId = process.env.GCLOUD_PROJECT;
+      core.info('Setting project Id from $GCLOUD_PROJECT');
+    } else if (!projectId) {
+      throw new Error('No project Id found. Set project Id in this action.');
+    }
+
+    this.parent = `projects/${projectId}/locations/${location}`;
+  }
+
+  /**
+   * Retrieves the auth client for authenticating requests.
+   *
+   * @returns JWT | Compute | UserRefreshClient.
+   */
+  async getAuthClient(): Promise<JWT | Compute | UserRefreshClient> {
+    if (!this.authClient) {
+      this.authClient = await this.auth.getClient();
+    }
+    return this.authClient;
+  }
+
+  /**
+   * Retrieves the auth client for authenticating requests.
+   *
+   * @returns JWT | Compute | UserRefreshClient.
+   */
+  async getToken(): Promise<string> {
+    const authClient = await this.getAuthClient();
+    const tokenResponse = await authClient.getAccessToken();
+    if (!tokenResponse.token) {
+      throw new Error('Unable to generate token.');
+    }
+    return tokenResponse.token;
+  }
+
+  /**
+   * Generates full resource name.
+   *
+   * @param cluster cluster name.
+   * @returns full resource name.
+   */
+  getResource(cluster: string): string {
+    return `${this.parent}/clusters/${cluster}`;
+  }
+
+  /**
+   * Retrieves a Cluster.
+   *
+   * @param fqn Cluster name
+   * @returns a Cluster object.
+   */
+  async getCluster(clusterName: string): Promise<ClusterResponse> {
+    const authClient = await this.getAuthClient();
+    const headers = await authClient.getRequestHeaders();
+    headers['User-Agent'] = this.userAgent;
+    const url = `${this.defaultEndpoint}/${this.getResource(clusterName)}`;
+    const resp = (await authClient.request({
+      url: url,
+      headers: headers,
+    })) as ClusterResponse;
+    return resp;
+  }
+
+  /**
+   * Create kubeconfig for cluster.
+   *
+   * @param authProvider boolean to use short lived OIDC token or GCP auth plugin in kubectl.
+   * @param useInternalIp boolean to use the internal IP address of the cluster endpoint.
+   * @returns kubeconfig
+   */
+  async createKubeConfig(
+    authProvider: string,
+    useInternalIp: string,
+    cluster: ClusterResponse,
+  ): Promise<string> {
+    const endpoint =
+      String(useInternalIp).toLowerCase() === 'true'
+        ? cluster.data.privateClusterConfig.privateEndpoint
+        : cluster.data.endpoint;
+    const auth =
+      String(authProvider).toLowerCase() === 'true'
+        ? 'user: {auth-provider: {name: gcp}}'
+        : `user: {token: ${await this.getToken()}}`;
+    const kubeConfig = `apiVersion: v1
```

Comment:
We may want to look into having this as an object then converting to YAML.

Reply:
Using types from `@kubernetes/client-node` ballooned the dist size to +10M, so I wrote types.